File: 2/Algo II/Practicos/Mis TADS/Implementaciones/TADConjunto.c
type conjunto = list;


fun is_empty (x: conjunto) ret b: bool {
	b = (x == []);
}

fun existe (x:conjunto, e: elem) ret b:bool {
	 while (x!i != e) {
		i++;
	 }
	 b = (x!i == e);
}

proc add (x:conjunto,e:elem) {
	i: nat = 1;
	if (is_empty(x)){
		p = e : x
	}
	else if (existe(x,e)) {
		break;
	}
	else {
		while (x!i < e) {
			i++;
		}
		x = take (i,x) : e : x
	}
}

proc union (x:conjunto, y:conjunto) {
	i:nat = 1;
	while (!(is_empty(x))) {
		add(y,x!i);
		i++;
	}
}

fun interseccion (x: conjunto, y:conjunto) ret z: conjunto {
	while (!is_empty(x)){
		if (existe(x!i,e)){
			add(x!i,z);
		}
	}
}

File: 2/Algo II/Practicos/Mis TADS/Especificaciones/TADConjunto.hs
module TADConjunto where
data Conjunto e = E | Add e (Conjunto e)
                deriving (Eq,Show)

existe :: Eq e => Conjunto e -> e ->  Bool
existe E _ = False
existe (Add e c) x = e == x || existe c x

interseccion :: Eq e => Conjunto e -> Conjunto e -> Conjunto e
interseccion E x = E
interseccion (Add e c) (Add s k) | existe (e k) = (Add e (interseccion c k)) 
                                 | otherwise = interseccion c k

union :: Conjunto e -> Conjunto e -> Conjunto e
union E x = x
union (Add e c) q = (Add e (union c q))

dif :: Eq e => Conjunto e -> Conjunto e -> Conjunto e
dif E x = x
dif (Add e c) (Add s k) | not (existe (e k)) = (Add e (dif c k)) 
                        | otherwise = dif c k